src/liballoc/collections/btree/navigate.rs

```diff
@@ -0,0 +1,258 @@
+    let (k, v) = kv.into_kv_mut();
+    (k, v) // coerce k from `&mut K` to `&K`
+}
+
+/// Moves the given leaf edge variable to the previous leaf and returns references to
+/// the key and value in between.
+/// Unsafe for two reasons:
+/// - the caller must ensure that the given leaf edge is not the first one in the tree;
+/// - using the updated handle may well invalidate the returned references.
+pub unsafe fn hop_to_next_back_unchecked_mut<'a, K, V>(
+    leaf_edge: &mut Handle<NodeRef<marker::Mut<'a>, K, V, marker::Leaf>, marker::Edge>,
+) -> (&'a K, &'a mut V) {
+    let (next_edge, kv) = next_back_unchecked_mut(ptr::read(leaf_edge));
+    *leaf_edge = next_edge;
+    // Doing the descend (and perhaps another move) invalidates the references
+    // returned by `into_kv_mut`, so we have to do this last.
+    let (k, v) = kv.into_kv_mut();
+    (k, v) // coerce k from `&mut K` to `&K`
+}
+
+/// Moves the given leaf edge variable to the next leaf edge and returns the key and value
+/// in between, while deallocating any node left behind.
+/// Unsafe for three reasons:
+/// - the caller must ensure that the given leaf edge is not the last one in the tree
+///   and is not a handle previously resulting from `hop_to_next_back_unchecked_deallocating`;
+/// - if the given handle is the last edge of a node, that node and possibly ancestors
+///   will be deallocated, while the reference to those nodes in the surviving ancestor
+///   is left dangling; thus further use of the leaf edge variable is dangerous.
+///   It is, however, safe to pass the leaf edge variable again to this function,
+///   if the two preconditions above hold.
+/// - using the updated handle in a safe way may well invalidate the returned references.
+pub unsafe fn hop_to_next_unchecked_deallocating<K, V>(
+    leaf_edge: &mut Handle<NodeRef<marker::Owned, K, V, marker::Leaf>, marker::Edge>,
+) -> (K, V) {
+    let (next_edge, k, v) = next_unchecked_deallocating(ptr::read(leaf_edge));
+    *leaf_edge = next_edge;
+    (k, v)
+}
+
+/// Moves the given leaf edge variable to the previous leaf edge and returns the key and value
+/// in between, while deallocating any node left behind.
+/// Unsafe for three reasons:
+/// - the caller must ensure that the given leaf edge is not the first one in the tree
+///   and is not a handle previously resulting from `hop_to_next_unchecked_deallocating`;
+/// - if the given handle is the first edge of a node, that node and possibly ancestors
+///   will be deallocated, while the reference to those nodes in the surviving ancestor
+///   is left dangling; thus further use of the leaf edge variable is dangerous.
+///   It is, however, safe to pass the leaf edge variable again to this function,
+///   if the two preconditions above hold.
+/// - using the updated handle in a safe way may well invalidate the returned references.
+pub unsafe fn hop_to_next_back_unchecked_deallocating<K, V>(
```

Comment:
(bikeshed) These names are quite a mouthful. What if we just made them all inherent methods on `Handle`? They can even overlap names since their borrow-type markers are different. Something like:

```rust
impl<'a, K, V> Handle<NodeRef<marker::Immut<'a>, K, V, marker::Leaf>, marker::Edge> {
    pub unsafe fn next_unchecked(&mut self) -> (&'a K, &'a V) { ...  }
    pub unsafe fn next_back_unchecked(&mut self) -> (&'a K, &'a V) { ...  }
}

impl<'a, K, V> Handle<NodeRef<marker::Mut<'a>, K, V, marker::Leaf>, marker::Edge> {
    pub unsafe fn next_unchecked(&mut self) -> (&'a K, &'a mut V) { ...  }
    pub unsafe fn next_back_unchecked(&mut self) -> (&'a K, &'a mut V) { ...  }
}

impl<K, V> Handle<NodeRef<marker::Owned, K, V, marker::Leaf>, marker::Edge> {
    pub unsafe fn next_unchecked(&mut self) -> (K, V) { ...  }
    pub unsafe fn next_back_unchecked(&mut self) -> (K, V) { ...  }
}
```

I think that will improve readability at the call sites too:

```diff
-            let (k, v) =
-                unsafe { navigate::hop_to_next_back_unchecked_deallocating(&mut self.back) };
-            Some((k, v))
+            Some(unsafe { self.back.next_back_unchecked() })
```

Reply:
I am happy either way, I don't know what's the usual or desired style is, but are you aware that `first_leaf_edge`/`last_leaf_edge` were standalone functions in map.rs, and that all `impl` blocks in map.rs are for types defined in itself? So I figured that in order to make for instance `first_leaf_edge` a method on `NodeRef`, like `first_edge` is, it would have to move to node.rs. Which would be both good (there is some similarity) and bad (node.rs is pretty fat already, though not as fat as map.rs). search.rs seemed most similar, and it has standalone functions.

Secondly, is it obvious that iterating over an owned tree deallocates it? I guess since it returns ownership of K and V, you have to realize the tree can't survive and the function has to be used in an IntoIterator. But if you're the owner of a tree and you somehow want to iterate through it (sending everyone a letter of resignation) without already destroying it?

Thirdly, if we do that, then do we remove the paper thin, private Range::next_back_unchecked function?

Reply:
Also, I stuck on the hop_on_ prefix specifically to distinguish them from the existing next_... functions. But since they're prefixed with `navigate::` when used, that was pointless.

Reply:
Naming is certainly subjective, which is why I prefixed this as a bikeshed comment. :slightly_smiling_face: 

> it would have to move to node.rs.

It does not have to move -- methods can be added from anywhere in the same crate. Maybe the original author of `first`/`last_leaf_edge` just didn't know you could do that. Such methods even get privacy according to where they are defined, rather than from the type definition.

These are all `pub` anyway, though not fully publicly reachable -- maybe they should be tightened to `pub(crate)` or even `pub(super)`. (That's a possible cleanup across most of `btree`.)

> Secondly, is it obvious that iterating over an owned tree deallocates it?

Seems ok to me, but that's a subjective point too.

> Thirdly, if we do that, then do we remove the paper thin, private Range::next_back_unchecked function?

There's still a _little_ value in discriminating the range's `front` or `back` for those that are calling `self.inner.next_unchecked` etc.

> Also, I stuck on the hop_on_ prefix specifically to distinguish them from the existing next_... functions. But since they're prefixed with `navigate::` when used, that was pointless.

Well, there aren't existing `next*` methods on `Handle`, at least, so I think it's ok. It's a nice symmetry IMO that outer `next` just calls the handle's `next_unchecked`, and `next_back` calls `next_back_unchecked`.

Reply:
> maybe they should be tightened to `pub(crate)` or even `pub(super)`.

Aren't they already effectively something like that because mod does not pub-lish module node?

Reply:
> Aren't they already effectively something like that because mod.rs does not pub-lish module node?

Right, they're not actually reachable, and the compiler knows that -- it will even tell you that with `RUSTFLAGS="-W unreachable_pub"` or `#![warn(unreachable_pub)]` in the code. That lint is allowed (silent) by default.

Reply:
Note that while rebasing, I noticed #66648 has added a `mut` to the return signature of `next_unchecked`. I'm doing the same in `next_back_unchecked`.

Reply:
Changing the `pub` visibility would be more about making sure they're never exposed by accident. But the compiler enforces stability attributes on truly public items in the standard library anyway, due to `#![feature(staged_api)]`, so I guess that doesn't matter.

Reply:
Also, added the same `is_empty` to `Range` as was added to `RangeMut` and narrowed down the unsafe sections.

Reply:
> These are all pub anyway, though not fully publicly reachable

Not sure what you mean with "not fully", but if they're not explicitly `pub` in navigate.rs, trying to call them in map.rs makes the compiler bark violently that they're private. So I'd rather say they're fully not public.